File: SoftProgDAO/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/pago/LineaComprobantePagoDAOImpl.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package pe.edu.pucp.inf30.softprog.daoimpl.pago;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import pe.edu.pucp.inf30.softprog.dao.pago.LineaComprobantePagoDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.TransaccionalBaseDAO;
import pe.edu.pucp.inf30.softprog.modelo.pago.LineaComprobantePago;
import pe.edu.pucp.inf30.softprog.modelo.venta.LineaOrdenCompra;

/**
 *
 * @author Cristhian Horacio
 */
public class LineaComprobantePagoDAOImpl extends TransaccionalBaseDAO<LineaComprobantePago> implements LineaComprobantePagoDAO{

    @Override
    protected PreparedStatement comandoCrear(Connection conn, LineaComprobantePago modelo) throws SQLException {
        String sql = "{CALL insertarLineaComprobantePago(?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idLineaComprobantePago", modelo.getId());
        cmd.setDouble("p_montoPagado", modelo.getMontoPagado());
        cmd.setDouble("p_montoImpuesto", modelo.getMontoImpuesto());
        cmd.setInt("p_activo", modelo.getActivo());
        cmd.setInt("p_ComprobantePago_idComprobante", modelo.getIdComprobantePago());
        cmd.setInt("p_codigo", modelo.getCodigo());
        cmd.setInt("p_cantidad", modelo.getCantidad());
        cmd.setDouble("p_subtotal", modelo.getSubTotal());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, LineaComprobantePago modelo) throws SQLException {
        String sql = "{CALL modificarLineaComprobantePago(?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idLineaComprobantePago", modelo.getId());
        cmd.setDouble("p_montoPagado", modelo.getMontoPagado());
        cmd.setDouble("p_montoImpuesto", modelo.getMontoImpuesto());
        cmd.setInt("p_activo", modelo.getActivo());
        cmd.setInt("p_ComprobantePago_idComprobante", modelo.getIdComprobantePago());
        cmd.setInt("p_codigo", modelo.getCodigo());
        cmd.setInt("p_cantidad", modelo.getCantidad());
        cmd.setDouble("p_subtotal", modelo.getSubTotal());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{CALL eliminarLineaComprobantePago(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idLineaComprobantePago", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{CALL buscarLineaComprobantePagoPorId(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        cmd.setInt("p_idLineaComprobantePago", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{CALL listarLineasComprobantePago()}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        return cmd;
    }

    @Override
    protected LineaComprobantePago mapearModelo(ResultSet rs) throws SQLException {
        LineaComprobantePago linea = new LineaComprobantePago();
        
        linea.setId(rs.getInt("idLineaComprobantePago"));
        linea.setMontoPagado(rs.getDouble("montoPagado"));
        linea.setMontoImpuesto(rs.getDouble("montoImpuesto"));
        linea.setActivoInt(rs.getInt("activo"));
        linea.setIdComprobantePago(rs.getInt("ComprobantePago_idComprobante"));
        linea.setCodigo(rs.getInt("codigo"));
        linea.setCantidad(rs.getInt("cantidad"));
        linea.setSubTotal(rs.getDouble("subtotal"));
        
        return linea;
    }

    @Override
    public List<LineaComprobantePago> listarPorIdComprobante(int id){
        return ejecutarComando(conn -> listarPorIdComprobante(conn, id));
    }
    
    protected PreparedStatement comandoListarPorIdComprobante(Connection conn, int id) throws SQLException{
        String sql = "{CALL listarLineasComprobantePagoPorIdComprobante(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idComprobante", id);
        
        return cmd;
    }
    
    @Override
    public List<LineaComprobantePago> listarPorIdComprobante(Connection conn, int id) {
        try(PreparedStatement cmd = this.comandoListarPorIdComprobante(conn, id)){
            ResultSet rs = cmd.executeQuery();
            
            List<LineaComprobantePago> modelos = new ArrayList<>();
            while(rs.next()){
                modelos.add(this.mapearModelo(rs));
            }
            
            return modelos;
        } catch (SQLException ex) {
            System.err.println("Error SQL: " + ex.getMessage());
            throw new RuntimeException(ex);
        }
    }
        
}

File: TrabajoAcademicoNet/SoftProgPersistencia/DAOImpl/DAOImplPago/LineaComprobantePagoDAOImpl.cs
using PUCP.Inf30.SoftProg.Modelo.PagoModelo;
using PUCP.Inf30.SoftProg.Modelo.PagoModelo.PagoUtils;
using PUCP.Inf30.SoftProg.Persistencia.DAO.DAOPago;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PUCP.Inf30.SoftProg.Persistencia.DAOImpl.DAOImplPago
{
    public class LineaComprobantePagoDAOImpl :
        TransaccionalBaseDAO<LineaComprobantePago>, ILineaComprobantePagoDAO
    {
        public List<LineaComprobantePago> ListarPorIdComprobante(int id)
        {
            throw new NotImplementedException();
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, LineaComprobantePago modelo)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarLineaComprobantePago";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idLineaComprobantePago", DbType.Int32, modelo.Id);
            this.AgregarParametroEntrada(cmd, "@p_montoPagado", DbType.DateTime, modelo.MontoPagado);
            this.AgregarParametroEntrada(cmd, "@p_montoImpuesto", DbType.Int32, modelo.MontoImpuesto);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Double, modelo.Activo);
            this.AgregarParametroEntrada(cmd, "@p_ComprobantePago_idComprobante", DbType.String, modelo.IdComprobantePago);
            this.AgregarParametroEntrada(cmd, "@p_codigo", DbType.Double, modelo.Codigo);
            this.AgregarParametroEntrada(cmd, "@p_cantidad", DbType.String, modelo.Cantidad);
            this.AgregarParametroEntrada(cmd, "@p_subtotal", DbType.String, modelo.SubTotal);

            return cmd;
        }

        protected override DbCommand ComandoCrear(DbConnection conn, LineaComprobantePago modelo)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarLineaComprobantePago";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idLineaComprobantePago", DbType.Int32, modelo.Id);
            this.AgregarParametroEntrada(cmd, "@p_montoPagado", DbType.DateTime, modelo.MontoPagado);
            this.AgregarParametroEntrada(cmd, "@p_montoImpuesto", DbType.Int32, modelo.MontoImpuesto);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Double, modelo.Activo);
            this.AgregarParametroEntrada(cmd, "@p_ComprobantePago_idComprobante", DbType.String, modelo.IdComprobantePago);
            this.AgregarParametroEntrada(cmd, "@p_codigo", DbType.Double, modelo.Codigo);
            this.AgregarParametroEntrada(cmd, "@p_cantidad", DbType.String, modelo.Cantidad);
            this.AgregarParametroEntrada(cmd, "@p_subtotal", DbType.String, modelo.SubTotal);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarLineaComprobantePago";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idLineaComprobantePago", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarLineaComprobantePago";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idLineaComprobantePago", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn)
        {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarLineaComprobantePago";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override LineaComprobantePago MapearModelo(DbDataReader reader)
        {
            return new LineaComprobantePago
            {
                Id = Convert.ToInt32(reader["idLineaComprobantePago"]),
                MontoPagado = Convert.ToDouble(reader["montoPagado"]),
                MontoImpuesto = Convert.ToDouble(reader["montoImpuesto"]),
                Activo = Convert.ToBoolean(reader["activo"]),
                IdComprobantePago = Convert.ToInt32(reader["ComprobantePago_idComprobante"]),
                Codigo = Convert.ToInt32(reader["codigo"]),
                Cantidad = Convert.ToInt32(reader["cantidad"]),
                SubTotal = Convert.ToDouble(reader["subtotal"])
            };
        }
        
    }
}
